### PR TITLE
Check winner or tie

### PR DESCRIPTION
**Pull request by GASANAMDavid**
[A game engine, needs to determine which player won for the game to be finished.](https://trello.com/c/4kb9Brhw/31-a-game-engine-need-to-determine-which-player-won-for-the-game-to-be-finished)

[A game engine needs to determine if the game is tied, for the game to be finished.](https://trello.com/c/5bKIKjxI/32-a-game-engine-needs-to-determine-if-the-game-is-tied-for-the-game-to-be-finished)

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
+  include BoardAsRows
   DENOTE_EMPTY = '-'
-  attr_reader :board, :num_of_blank_spaces, :blank_positions
+  attr_reader :board, :num_of_played_positions, :blank_positions
 
   def initialize(size)
     @board = Array.new(size) { Array.new(size) { DENOTE_EMPTY } }
-    @num_of_blank_spaces = size * size
+    @num_of_played_positions = 0
     @blank_positions = make_free_positions_hash(size.times.to_a.product(size.times.to_a))
```
Comment (naush) [inline lib/board.rb]: It looks like `empty`, `blank`, and `free` may all mean the same thing. Let's pick one and be consistent throughout the code base.

File: lib/board.rb [hunk 2]
```diff
@@ -27,4 +30,22 @@ def make_free_positions_hash(blanks)
     end
     hash
   end
+
+  def board_state(symbol)
+    return nil if num_of_played_positions < board.size * 2 - 1
```
Comment (naush) [inline lib/board.rb]: `num_of_played_positions < board.size * 2 - 1` might be a good place to introduce an explaining variable.

Comment (naush) [reply lib/board.rb]: What would happen if the game is over before all the positions are taken? For example,

```
o o +
x x x
+ + +
```

File: lib/board.rb [hunk 3]
```diff
@@ -27,4 +30,22 @@ def make_free_positions_hash(blanks)
...
+    if check_rows.any? { |row| check_uniqueness(row, symbol) }
+      'Winner'
+    elsif num_of_played_positions == board_size * board_size
+      'Tie'
+    end
+  end
+
+  private
+
+  def check_uniqueness(row, mark)
```
Comment (naush) [inline lib/board.rb]: Another method name I may consider is `win?` since it is a little more intentional.

File: spec/board_spec.rb [hunk 4]
```diff
@@ -8,20 +8,30 @@
...
-    player_board.print_board(output)
-    expect(output.string).to include("[\"-\", \"-\", \"-\"]\n")
-  end
-
   it "adds the player's symbol to the board's specified index " do
     player_board.apply_move('X', { row: 0, col: 0 })
     expect(player_board.board).to eq([%w[X - -], %w[- - -], %w[- - -]])
   end
+  describe '#board_state' do
+    context 'when the game is won' do
```
Comment (naush) [inline spec/board_spec.rb:16]: I would also include cases that check when the game is not over yet or tied for completeness.

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -47,19 +47,19 @@
...
-    $stderr = $origin_stderr
-    $origin_stdout = nil
-    $origin_stderr = nil
-  end
+  # config.after(:all) do
+  #   $stdout = $origin_stdout
+  #   $stderr = $origin_stderr
+  #   $origin_stdout = nil
+  #   $origin_stderr = nil
+  # end
```
Comment (naush) [inline spec/spec_helper.rb]: Feel free to remove any commented out code. Remember we can always get them back from Git history!

File: spec/board_spec.rb (hunk 4 above)
Comment (GASANAMDavid) [reply spec/board_spec.rb:16]: Resolved at 5fbd8db67c30a860f9e237172e7fd1152cb36f01

File: lib/board.rb (hunk 1 above)
Comment (GASANAMDavid) [reply lib/board.rb]: Resolved at 5fbd8db67c30a860f9e237172e7fd1152cb36f01

File: lib/board.rb (hunk 2 above)
Comment (GASANAMDavid) [reply lib/board.rb]: Resolved at 4366f41b26aec5c9cbf1f087275c1ec47033cacb

File: spec/spec_helper.rb (hunk 5 above)
Comment (GASANAMDavid) [reply spec/spec_helper.rb]: Resolve at 4366f41b26aec5c9cbf1f087275c1ec47033cacb

Comment (GASANAMDavid) [reply spec/spec_helper.rb]: Resolved at 8c52707d7bd8ae4d77b90c08bc7555fcc3456aac

File: lib/board.rb (hunk 3 above)
Comment (GASANAMDavid) [reply lib/board.rb]: REsolved at 9dbbabed8c5453b180590dc36e5ca3e44f5c06c8

File: lib/board.rb (hunk 1 above)
Comment (GASANAMDavid) [reply lib/board.rb]: Resolved at 9dbbabed8c5453b180590dc36e5ca3e44f5c06c8

Comment (GASANAMDavid) [reply lib/board.rb]: Resolved at 2d157729000732e9b53f581d1f5213a341c3fed0

**Pull request closed without merging**